File: backend/routers/game/chat_mode.py
"""
Chat mode route handlers.

Handles /chat and /end commands for free-form NPC conversations.
"""

import asyncio
import logging
from typing import Optional

import crud
import models
import schemas
from domain.entities.agent import is_chat_summarizer
from domain.entities.agent_config import AgentConfigData
from domain.value_objects.contexts import AgentResponseContext
from domain.value_objects.enums import MessageRole
from domain.value_objects.task_identifier import TaskIdentifier
from orchestration import get_chat_mode_orchestrator
from sdk import AgentManager
from sdk.agent.options_builder import build_agent_options
from services.agent_config_service import AgentConfigService
from services.prompt_builder import build_system_prompt
from sqlalchemy.ext.asyncio import AsyncSession
from utils.images import compress_image_base64

logger = logging.getLogger("ChatModeRoutes")


async def _warm_chat_summarizer(room_id: int, agent_manager: AgentManager) -> None:
    """
    Pre-warm the Chat_Summarizer client by creating it in the client pool.

    This reduces latency when exiting chat mode (/end) by having the SDK client
    already connected and ready. Called as a background task when entering chat mode.

    Args:
        room_id: Room ID for the client pool key
        agent_manager: AgentManager instance with client pool
    """
    from database import get_db as get_db_generator

    async for db in get_db_generator():
        try:
            # Get the Chat_Summarizer agent
            summarizer = await _get_chat_summarizer_agent(db)
            if not summarizer:
                logger.warning("Cannot warm Chat_Summarizer: agent not found")
                return

            # Load agent config from filesystem
            config_data = AgentConfigData()
            if summarizer.config_file:
                loaded_config = AgentConfigService.load_agent_config(summarizer.config_file)
                if loaded_config:
                    config_data = loaded_config

            # Build system prompt
            system_prompt = build_system_prompt(summarizer.name, config_data)

            # Build AgentResponseContext (user_message not used for warming)
            task_id = TaskIdentifier(room_id=room_id, agent_id=summarizer.id)
            context = AgentResponseContext(
                system_prompt=system_prompt,
                user_message="",  # Not used for warming
                agent_name=summarizer.name,
                config=config_data,
                room_id=room_id,
                agent_id=summarizer.id,
                group_name=summarizer.group,
                task_id=task_id,
            )

            # Build agent options (same logic as generate_sdk_response)
            options, config_hash = build_agent_options(context, system_prompt, [])

            # Pre-create the client in the pool
            _pooled, is_new, _lock = await agent_manager.client_pool.get_or_create(task_id, options, config_hash)

            if is_new:
                logger.info(f"Chat_Summarizer client warmed for room {room_id}")
            else:
                logger.debug(f"Chat_Summarizer client already warm for room {room_id}")

        except Exception as e:
            logger.warning(f"Failed to warm Chat_Summarizer client: {e}")
        finally:
            break  # Only use first session


async def handle_chat_command(
    db: AsyncSession,
    world_id: int,
    player_state: models.PlayerState,
    room_id: int,
    _world: models.World,
    agent_manager: AgentManager,
) -> dict:
    """
    Enter chat mode.

    Args:
        db: Database session
        world_id: World ID
        player_state: Current player state
        room_id: Target room ID
        world: World model
        agent_manager: AgentManager instance for warming Chat_Summarizer

    Returns:
        Response dict with status
    """
    if player_state.is_chat_mode:
        return {
            "status": "already_in_chat_mode",
            "message": "You are already in chat mode. Type /end to return to gameplay.",
        }

    # Get latest message ID as start marker
    messages = await crud.get_recent_messages(db, room_id, limit=1)
    start_message_id = messages[-1].id if messages else 0

    # Enter chat mode (returns chat_session_id)
    chat_session_id = await crud.enter_chat_mode(db, world_id, start_message_id)
    if chat_session_id is None:
        return {
            "status": "error",
            "message": "Failed to enter chat mode.",
        }

    # Send system message (this message is hidden from polling, but stored for context)
    # System messages in chat mode also get the chat_session_id
    await crud.create_message(
        db,
        room_id,
        schemas.MessageCreate(
            content="[Chat mode started. You can now freely converse with NPCs. Type /end to return to gameplay.]",
            role=MessageRole.USER,
            participant_type="system",
            participant_name="System",
            chat_session_id=chat_session_id,
        ),
        update_room_activity=True,
    )

    # Warm the Chat_Summarizer client in background (reduces /end latency)
    asyncio.create_task(_warm_chat_summarizer(room_id, agent_manager))

    logger.info(
        f"Entered chat mode for world {world_id}, start_message_id={start_message_id}, chat_session_id={chat_session_id}"
    )

    return {
        "status": "chat_mode_started",
        "message": "Chat mode started. You can now freely converse with NPCs. Type /end to return to gameplay.",
    }


async def handle_chat_mode_action(
    db: AsyncSession,
    world_id: int,
    player_state: models.PlayerState,
    room_id: int,
    text: str,
    agent_manager: AgentManager,
    world: models.World,
    location_id: int,
    image_data: Optional[str] = None,
    image_media_type: Optional[str] = None,
) -> dict:
    """
    Handle player message in chat mode.

    Triggers NPC responses without Action Manager or Narrator.

    Args:
        db: Database session
        world_id: World ID
        player_state: Current player state
        room_id: Target room ID
        text: Player's message text
        agent_manager: AgentManager instance
        world: World model
        location_id: Current location ID

    Returns:
        Response dict with status
    """
    # Get the current chat session ID from player state
    chat_session_id = player_state.chat_session_id

    # Compress image if present
    compressed_image_data = image_data
    compressed_image_media_type = image_media_type
    if compressed_image_data and compressed_image_media_type:
        try:
            logger.info(f"Compressing image for chat mode in world {world_id}")
            compressed_data, compressed_media_type = compress_image_base64(
                compressed_image_data, compressed_image_media_type
            )
            original_size = len(compressed_image_data)
            compressed_size = len(compressed_data)
            compression_ratio = (1 - compressed_size / original_size) * 100 if original_size > 0 else 0
            logger.info(
                f"Image compressed: {original_size} -> {compressed_size} bytes ({compression_ratio:.1f}% reduction)"
            )
            compressed_image_data = compressed_data
            compressed_image_media_type = compressed_media_type
        except Exception as e:
            logger.warning(f"Image compression failed, using original: {e}")

    # Save user message to room with chat_session_id
    message = schemas.MessageCreate(
        content=text,
        role=MessageRole.USER,
        participant_type="user",
        chat_session_id=chat_session_id,
        image_data=compressed_image_data,
        image_media_type=compressed_image_media_type,
    )
    await crud.create_message(db, room_id, message, update_room_activity=True)

    # Trigger chat mode NPC responses in background
    async def trigger_chat_responses():
        """Background task to trigger chat mode NPC responses."""
        from database import get_db as get_db_generator

        async for task_db in get_db_generator():
            try:
                chat_orchestrator = get_chat_mode_orchestrator()
                await chat_orchestrator.handle_chat_message(
                    db=task_db,
                    room_id=room_id,
                    message_text=text,
                    agent_manager=agent_manager,
                    world_id=world_id,
                    world_name=world.name,
                    location_id=location_id,
                    chat_session_id=chat_session_id,
                )
            except Exception as e:
                logger.error(f"Error triggering chat mode responses: {e}")
                import traceback

                traceback.print_exc()
            finally:
                break  # Only use first session

    asyncio.create_task(trigger_chat_responses())
    logger.info(f"Chat mode message submitted for world {world_id}: {text[:50]}...")

    return {
        "status": "processing",
        "message": "Message received, NPCs are responding...",
    }


async def handle_end_command(
    db: AsyncSession,
    world_id: int,
    player_state: models.PlayerState,
    room_id: int,
    agent_manager: AgentManager,
    world: models.World,
) -> dict:
    """
    Exit chat mode and summarize the conversation.

    If no chat interaction happened, exits silently without summarization.

    Args:
        db: Database session
        world_id: World ID
        player_state: Current player state
        room_id: Target room ID
        agent_manager: AgentManager instance
        world: World model

    Returns:
        Response dict with status
    """
    if not player_state.is_chat_mode:
        return {
            "status": "not_in_chat_mode",
            "message": "You are not in chat mode.",
        }

    # Get the chat session ID before exiting
    chat_session_id = player_state.chat_session_id

    # Check if there was any actual chat interaction (non-system messages)
    has_chat_interaction = False
    if chat_session_id:
        messages = await crud.get_chat_session_messages(db, room_id, chat_session_id, limit=10)
        # Check for non-system messages
        for m in messages:
            if m.participant_type != "system":
                has_chat_interaction = True
                break

    # Exit chat mode and get start message ID and chat session ID
    exit_result = await crud.exit_chat_mode(db, world_id)
    if exit_result is None:
        return {
            "status": "error",
            "message": "Failed to exit chat mode.",
        }

    _start_message_id, returned_chat_session_id = exit_result

    # If no chat interaction, exit silently without summarization
    if not has_chat_interaction:
        logger.info(f"Exited chat mode for world {world_id} with no interaction, skipping summarizer")
        return {
            "status": "chat_mode_ended",
            "message": "Exited chat mode.",
        }

    # Send system message for end of chat mode (no chat_session_id since we're ending)
    await crud.create_message(
        db,
        room_id,
        schemas.MessageCreate(
            content="[Chat mode ended. Returning to gameplay...]",
            role=MessageRole.USER,
            participant_type="system",
            participant_name="System",
        ),
        update_room_activity=True,
    )

    # Background task: summarize the conversation and continue gameplay
    asyncio.create_task(
        _summarize_and_continue(
            world_id=world_id,
            _world_name=world.name,
            room_id=room_id,
            chat_session_id=returned_chat_session_id,
            agent_manager=agent_manager,
            user_name=world.user_name or "The player",
            world_genre=world.genre,
        )
    )

    logger.info(f"Exited chat mode for world {world_id}, summarizing conversation...")

    return {
        "status": "processing",
        "message": "Returning to gameplay...",
    }


async def _get_chat_summarizer_agent(db: AsyncSession) -> Optional[models.Agent]:
    """
    Get the Chat_Summarizer agent from the database.

    Returns:
        Agent model or None if not found
    """
    # Find Chat_Summarizer in gameplay agents
    from sqlalchemy import select

    result = await db.execute(select(models.Agent).where(models.Agent.group == "gameplay"))
    agents = result.scalars().all()

    for agent in agents:
        if is_chat_summarizer(agent.name):
            return agent

    logger.warning("Chat_Summarizer agent not found in database")
    return None


async def _generate_ai_summary(
    db: AsyncSession,
    agent_manager: AgentManager,
    room_id: int,
    conversation_text: str,
    participants: set[str],
) -> Optional[str]:
    """
    Generate an AI summary of the conversation using the Chat_Summarizer agent.

    Args:
        db: Database session
        agent_manager: AgentManager instance
        room_id: Room ID for task tracking
        conversation_text: Formatted conversation transcript
        participants: Set of NPC names who participated

    Returns:
        AI-generated summary text, or None if summarizer unavailable
    """
    # Get the Chat_Summarizer agent
    summarizer = await _get_chat_summarizer_agent(db)
    if not summarizer:
        return None

    try:
        # Load agent config from filesystem
        config_data = AgentConfigData()
        if summarizer.config_file:
            loaded_config = AgentConfigService.load_agent_config(summarizer.config_file)
            if loaded_config:
                config_data = loaded_config

        # Build system prompt
        system_prompt = build_system_prompt(summarizer.name, config_data)

        # Build the user message with conversation context
        participant_list = ", ".join(participants) if participants else "NPCs"
        user_message = f"""Please summarize the following conversation between the player and {participant_list}.

## Conversation Transcript
{conversation_text}

## Instructions
Create a concise 2-4 sentence summary focusing on:
- Key topics discussed
- Important information exchanged
- Any agreements or outcomes
- Relationship changes (if any)

Write in past tense, third person (e.g., "The player discussed...", "They agreed...")."""

        # Build AgentResponseContext
        task_id = TaskIdentifier(room_id=room_id, agent_id=summarizer.id)
        context = AgentResponseContext(
            system_prompt=system_prompt,
            user_message=user_message,
            agent_name=summarizer.name,
            config=config_data,
            room_id=room_id,
            agent_id=summarizer.id,
            group_name=summarizer.group,
            task_id=task_id,
        )

        # Generate summary via AgentManager
        response_text = ""
        async for event in agent_manager.generate_sdk_response(context):
            if event.get("type") == "content_delta":
                response_text += event.get("delta", "")
            elif event.get("type") == "stream_end":
                # Use final response_text from stream_end if available
                if event.get("response_text"):
                    response_text = event["response_text"]

        logger.info(f"Chat_Summarizer generated summary: {response_text[:100]}...")
        return response_text.strip() if response_text else None

    except Exception as e:
        logger.error(f"Error generating AI summary: {e}")
        import traceback

        traceback.print_exc()
        return None


async def _summarize_and_continue(
    world_id: int,
    _world_name: str,
    room_id: int,
    chat_session_id: int,
    agent_manager: AgentManager,
    user_name: str,
    world_genre: Optional[str],
) -> None:
    """
    Background task to summarize chat mode conversation and continue gameplay.

    1. Get messages from chat mode session
    2. Generate AI summary using Chat_Summarizer agent
    3. Pass summary to Action Manager -> Narrator via TRPG orchestrator

    If no chat interaction happened, skips summarization entirely.
    """
    from database import get_db as get_db_generator
    from orchestration import get_trpg_orchestrator

    trpg_orchestrator = get_trpg_orchestrator()

    async for db in get_db_generator():
        try:
            # Ensure gameplay agents are in the location room
            # (they might be missing if location was created before agents were seeded)
            from crud.worlds import add_gameplay_agents_to_room

            await add_gameplay_agents_to_room(db, room_id)

            # Get messages from the chat session
            messages = await crud.get_chat_session_messages(db, room_id, chat_session_id, limit=100)

            if not messages:
                logger.info(f"No messages to summarize for world {world_id}, skipping summarizer")
                return

            # Format messages as a conversation transcript
            conversation_messages = []
            participants: set[str] = set()
            for m in messages:
                if m.participant_type == "system":
                    continue  # Skip system messages
                if m.role == MessageRole.USER:
                    name = m.participant_name or user_name
                    conversation_messages.append(f"{name}: {m.content}")
                else:
                    name = m.agent.name if m.agent else "Unknown"
                    conversation_messages.append(f"{name}: {m.content}")
                    if m.agent and m.agent.name:
                        participants.add(m.agent.name)

            if not conversation_messages:
                logger.info(f"No conversation content to summarize for world {world_id}, skipping summarizer")
                return

            # Only mark as processing AFTER confirming there's content to summarize
            trpg_orchestrator.set_sub_agent_active(room_id, "Chat_Summarizer", "Summarizing conversation...")

            # Format conversation for summarizer
            conversation_text = "\n".join(conversation_messages)

            # Generate AI summary using Chat_Summarizer agent
            ai_summary = await _generate_ai_summary(
                db=db,
                agent_manager=agent_manager,
                room_id=room_id,
                conversation_text=conversation_text,
                participants=participants,
            )

            # Fall back to simple summary if AI summarizer unavailable
            if ai_summary:
                summary_text = ai_summary
            else:
                # Fallback: use last few exchanges
                participant_list = ", ".join(participants) if participants else "NPCs"
                recent_exchanges = conversation_messages[-6:]
                summary_text = f"[End of conversation with {participant_list}]\nRecent exchanges:\n" + "\n".join(
                    recent_exchanges
                )

            logger.info(f"Chat mode ended, passing to gameplay: {summary_text[:100]}...")

            # Clear sub-agent status before continuing to gameplay agents
            trpg_orchestrator.set_sub_agent_inactive(room_id)

            # Now trigger TRPG flow with the summary as the action
            # The Action Manager will interpret the summary and Narrator will describe
            # Re-fetch world in this session
            task_world = await crud.get_world(db, world_id)
            if task_world:
                # Create a synthetic action that summarizes what happened
                summary_action = f"[Conversation Summary] {summary_text}"

                await trpg_orchestrator.handle_player_action(
                    db=db,
                    room_id=room_id,
                    action_text=summary_action,
                    agent_manager=agent_manager,
                    world=task_world,
                )

        except Exception as e:
            logger.error(f"Error in summarize_and_continue: {e}")
            import traceback

            traceback.print_exc()
        finally:
            # Ensure sub-agent status is cleared even on error
            trpg_orchestrator.set_sub_agent_inactive(room_id)
            break  # Only use first session

File: backend/src/http/routes/game/chat-mode.ts
/**
 * Chat mode: a side conversation with the NPCs at the player's location, with
 * no Action Manager, no narration, no turn counter and no stat changes. Not a
 * router — `/chat` and `/end` arrive as slash commands inside a normal action
 * body, and `actions.ts` dispatches here. The status is always 200, failures
 * included; the frontend renders those as a system line.
 *
 * The separation mechanism is `messages.chat_session_id`, set on chat-mode
 * messages and NULL on gameplay ones. `/end` hands the transcript to a
 * Chat_Summarizer and feeds its summary back into the tape as a synthetic
 * player action, which is why it can answer `processing`.
 */

import { eq } from 'drizzle-orm'

import { getRoomAgentSession, updateRoomAgentSession } from '@/crud/sessions'
import { createMessage, getChatSessionMessages, getRecentMessages } from '@/crud/messages'
import { enterChatMode, exitChatMode } from '@/crud/player-state'
import { addGameplayAgentsToRoom, getWorld } from '@/crud/worlds'
import type { Db } from '@/db'
import { agents, type Agent, type PlayerState, type World } from '@/db/schema'
import { isChatSummarizer } from '@/domain/agent'
import { getSettings } from '@/config/settings'
import { getLogger } from '@/infrastructure/logging/logger'
import type { AgentOptionsInput } from '@/sdk/agent/options-builder'
import { buildAgentOptions, optionsFingerprint } from '@/sdk/agent/options-builder'
import { TurnRunner } from '@/sdk/agent/turn-runner'
import { parseAgentConfig } from '@/sdk/parsing/agent-config'
import { buildSystemPrompt } from '@/services/prompt-builder'
import type { AppState } from '@/http/state'
import { startBackground, tryCompressImage } from './shared'

const logger = getLogger('ChatModeRoutes')

/** Every chat-mode handler answers 200 with one of these. */
export interface ChatModeResult {
  status: string
  message: string
}

// The start marker — the id of the room's newest message, or `0` — is where the
// frontend resumes the gameplay transcript, so it must be captured *before* the
// "[Chat mode started...]" message is written.
export function handleChatCommand(
  state: AppState,
  worldId: number,
  playerState: PlayerState,
  roomId: number,
): ChatModeResult {
  if (playerState.isChatMode) {
    return {
      status: 'already_in_chat_mode',
      message: 'You are already in chat mode. Type /end to return to gameplay.',
    }
  }

  const recent = getRecentMessages(state.db, roomId, 1)
  const startMessageId = recent[recent.length - 1]?.id ?? 0

  const chatSessionId = enterChatMode(state.db, worldId, startMessageId)
  if (chatSessionId === null) {
    return { status: 'error', message: 'Failed to enter chat mode.' }
  }

  // Tagged with the session id: hidden from the poll, but part of the
  // conversation the summarizer reads.
  createMessage(state.db, roomId, {
    content:
      '[Chat mode started. You can now freely converse with NPCs. Type /end to return to gameplay.]',
    role: 'user',
    participantType: 'system',
    participantName: 'System',
    chatSessionId,
  })

  startBackground(() => warmChatSummarizer(state, roomId), {
    name: `warm_chat_summarizer:room=${roomId}`,
  })

  logger.info(
    `Entered chat mode for world ${worldId}, start_message_id=${startMessageId}, ` +
      `chat_session_id=${chatSessionId}`,
  )

  return {
    status: 'chat_mode_started',
    message: 'Chat mode started. You can now freely converse with NPCs. Type /end to return to gameplay.',
  }
}

export interface ChatModeActionInput {
  worldId: number
  playerState: PlayerState
  roomId: number
  text: string
  world: World
  imageData?: string | null
  imageMediaType?: string | null
}

// Store the message, then let the NPCs answer. No action history, no turn
// increment, no game-time snapshot: chat mode does not advance the game.
export async function handleChatModeAction(
  state: AppState,
  input: ChatModeActionInput,
): Promise<ChatModeResult> {
  const chatSessionId = input.playerState.chatSessionId

  const image = await tryCompressImage(
    input.imageData,
    input.imageMediaType,
    `chat mode in world ${input.worldId}`,
  )

  createMessage(state.db, input.roomId, {
    content: input.text,
    role: 'user',
    participantType: 'user',
    chatSessionId,
    imageData: image.imageData,
    imageMediaType: image.imageMediaType,
  })

  startBackground(
    async () => {
      await state.orchestrator.handleChatMessage({
        world: input.world,
        roomId: input.roomId,
        action: input.text,
        chatSessionId,
      })
    },
    { name: `trigger_chat_responses:world=${input.worldId}` },
  )

  logger.info(`Chat mode message submitted for world ${input.worldId}: ${input.text.slice(0, 50)}...`)

  return { status: 'processing', message: 'Message received, NPCs are responding...' }
}

/**
 * Exit chat mode, summarizing unless there was no conversation. The interaction
 * check runs *before* the exit, which clears the `chat_session_id` it needs. A
 * session holding only the "[Chat mode started...]" line exits silently.
 */
export function handleEndCommand(
  state: AppState,
  worldId: number,
  playerState: PlayerState,
  roomId: number,
  world: World,
): ChatModeResult {
  if (!playerState.isChatMode) {
    return { status: 'not_in_chat_mode', message: 'You are not in chat mode.' }
  }

  const chatSessionId = playerState.chatSessionId
  let hasChatInteraction = false
  if (chatSessionId) {
    hasChatInteraction = getChatSessionMessages(state.db, roomId, chatSessionId, 10).some(
      (m) => m.participantType !== 'system',
    )
  }

  const exitResult = exitChatMode(state.db, worldId)
  if (exitResult === null) {
    return { status: 'error', message: 'Failed to exit chat mode.' }
  }

  if (!hasChatInteraction) {
    logger.info(`Exited chat mode for world ${worldId} with no interaction, skipping summarizer`)
    return { status: 'chat_mode_ended', message: 'Exited chat mode.' }
  }

  // No `chat_session_id`: tagging it would pull the closing line into the
  // transcript the summarizer is already reading.
  createMessage(state.db, roomId, {
    content: '[Chat mode ended. Returning to gameplay...]',
    role: 'user',
    participantType: 'system',
    participantName: 'System',
  })

  const returnedSessionId = exitResult.chatSessionId
  if (returnedSessionId !== null) {
    startBackground(
      () =>
        summarizeAndContinue(state, {
          worldId,
          roomId,
          chatSessionId: returnedSessionId,
          userName: world.userName || 'The player',
        }),
      { name: `summarize_and_continue:world=${worldId}` },
    )
  }

  logger.info(`Exited chat mode for world ${worldId}, summarizing conversation...`)
  return { status: 'processing', message: 'Returning to gameplay...' }
}

// By pattern, not exact name: the folder may be `Chat_Summarizer`,
// `chat summarizer` or `ChatSummarizer`.
function getChatSummarizerAgent(db: Db): Agent | null {
  const found = db
    .select()
    .from(agents)
    .where(eq(agents.group, 'gameplay'))
    .all()
    .find((agent) => isChatSummarizer(agent.name))

  if (!found) logger.warning('Chat_Summarizer agent not found in database')
  return found ?? null
}

function chatSummarizerOptions(state: AppState, summarizer: Agent, roomId: number): AgentOptionsInput {
  const config = summarizer.configFile ? parseAgentConfig(summarizer.configFile) : null
  const systemPrompt = config
    ? buildSystemPrompt(summarizer.name, config)
    : summarizer.systemPrompt

  return {
    systemPrompt,
    // No tools: the summarizer only produces prose, and the gameplay tools would
    // let a summary mutate the world it is describing.
    mcpServers: {},
    toolNames: [],
    resume: getRoomAgentSession(state.db, roomId, summarizer.id) ?? undefined,
    useSonnet: getSettings().useSonnet,
  }
}

// Pure latency work: `/end` otherwise waits on a cold subprocess start.
// Failures are swallowed — the session is then opened on demand.
async function warmChatSummarizer(state: AppState, roomId: number): Promise<void> {
  try {
    const summarizer = getChatSummarizerAgent(state.db)
    if (!summarizer) {
      logger.warning('Cannot warm Chat_Summarizer: agent not found')
      return
    }

    const input = chatSummarizerOptions(state, summarizer, roomId)
    await state.pool.acquire(
      { roomId, agentId: summarizer.id },
      buildAgentOptions(input),
      optionsFingerprint(input),
    )
    logger.info(`Chat_Summarizer client warmed for room ${roomId}`)
  } catch (error) {
    logger.warning(`Failed to warm Chat_Summarizer client: ${String(error)}`)
  }
}

async function generateAiSummary(
  state: AppState,
  roomId: number,
  conversationText: string,
  participants: Set<string>,
): Promise<string | null> {
  const summarizer = getChatSummarizerAgent(state.db)
  if (!summarizer) return null

  try {
    const options = chatSummarizerOptions(state, summarizer, roomId)
    const participantList = participants.size > 0 ? [...participants].join(', ') : 'NPCs'
    const userMessage = `Please summarize the following conversation between the player and ${participantList}.

## Conversation Transcript
${conversationText}

## Instructions
Create a concise 2-4 sentence summary focusing on:
- Key topics discussed
- Important information exchanged
- Any agreements or outcomes
- Relationship changes (if any)

Write in past tense, third person (e.g., "The player discussed...", "They agreed...").`

    let responseText = ''
    for await (const event of new TurnRunner(state.pool).run({
      roomId,
      agentId: summarizer.id,
      agentName: summarizer.name,
      content: userMessage,
      options,
      hidden: true,
    })) {
      if (event.type === 'content_delta') {
        responseText += event.delta
      } else if (event.type === 'stream_end') {
        // Written back here too, or the next `/end` resumes a stale session.
        if (event.sessionId && event.sessionId !== options.resume) {
          updateRoomAgentSession(state.db, roomId, summarizer.id, event.sessionId)
        }
        if (event.responseText) responseText = event.responseText
      }
    }

    logger.info(`Chat_Summarizer generated summary: ${responseText.slice(0, 100)}...`)
    return responseText.trim() || null
  } catch (error) {
    logger.exception('Error generating AI summary', error)
    return null
  }
}

interface SummarizeInput {
  worldId: number
  roomId: number
  chatSessionId: number
  userName: string
}

/**
 * Summarize the finished chat session and feed it back into gameplay. Three
 * orderings are load-bearing: gameplay agents are re-added first (a location
 * seeded before they existed has no Action Manager for the turn below); the
 * "sub-agent busy" indicator is set only once the transcript is known non-empty;
 * and it is cleared *before* the gameplay turn. With no summarizer available the
 * last six exchanges are pasted in verbatim.
 */
async function summarizeAndContinue(state: AppState, input: SummarizeInput): Promise<void> {
  const { roomId, worldId } = input

  try {
    addGameplayAgentsToRoom(state.db, roomId)

    const messages = getChatSessionMessages(state.db, roomId, input.chatSessionId, 100)
    if (messages.length === 0) {
      logger.info(`No messages to summarize for world ${worldId}, skipping summarizer`)
      return
    }

    const conversation: string[] = []
    const participants = new Set<string>()
    for (const message of messages) {
      if (message.participantType === 'system') continue
      if (message.role === 'user') {
        conversation.push(`${message.participantName || input.userName}: ${message.content}`)
      } else {
        const name = message.agent?.name ?? 'Unknown'
        conversation.push(`${name}: ${message.content}`)
        if (message.agent?.name) participants.add(message.agent.name)
      }
    }

    if (conversation.length === 0) {
      logger.info(`No conversation content to summarize for world ${worldId}, skipping summarizer`)
      return
    }

    state.orchestrator.setSubAgentActive(roomId, 'Chat_Summarizer', 'Summarizing conversation...')

    const aiSummary = await generateAiSummary(state, roomId, conversation.join('\n'), participants)

    const summaryText =
      aiSummary ??
      `[End of conversation with ${participants.size > 0 ? [...participants].join(', ') : 'NPCs'}]\n` +
        `Recent exchanges:\n${conversation.slice(-6).join('\n')}`

    logger.info(`Chat mode ended, passing to gameplay: ${summaryText.slice(0, 100)}...`)

    state.orchestrator.setSubAgentInactive(roomId)

    const world = getWorld(state.db, worldId)
    if (world) {
      await state.orchestrator.handlePlayerAction({
        world,
        roomId,
        action: `[Conversation Summary] ${summaryText}`,
      })
    }
  } catch (error) {
    logger.exception('Error in summarize_and_continue', error)
  } finally {
    state.orchestrator.setSubAgentInactive(roomId)
  }
}
